### PR TITLE
feature(i18n): allow option to force language with an url parameter

### LINKED ISSUE #3732: Add ability to force language via url parameter (Trac #3732)
_Original ticket http://trac.elgg.org/ticket/3732 on 41633680-03-19 by ewinslow, assigned to unknown._

Elgg version: 1.7

Could have sworn I already reported this... Oh well.

&lang=en should force English, for example. One way this comes majorly in handy is that it's easier to test other languages. Instead of changing your usersettings you can just change the url.

### PR DESCRIPTION
**Pull request by jdalsem**
fixes #3732

### CODE REVIEW

**Comment by juho-jaakkola** [comment]
Why ```__elgg_lang``` instead of just ```lang```?

File: engine/classes/Elgg/I18n/Translator.php [hunk 1]
```diff
@@ -147,6 +147,11 @@ function getCurrentLanguage() {
 	 * @return string The language code (eg "en") or false if not set
 	 */
 	function getLanguage() {
+		$url_lang = get_input('__elgg_lang');
```
Comment (ewinslow) [inline engine/classes/Elgg/I18n/Translator.php]: Use an api from `_elgg_services()` here instead of `get_input`.

**Comment by ewinslow** [comment]
Agree that __elgg_lang is a little awkward. This is something that end users will see.

I was thinking "locale", since you could have en-US, or equivalent, which is not just a language strictly speaking, but I think I'm probably the only one that would be bothered by this :)

**Comment by ewinslow** [comment]
But thanks for the change! This will be great!

**Comment by ewinslow** [comment]
Any test stubs you can add will be appreciated. Usually just taking the commit summary and naming a test method after it is sufficient.

**Comment by jdalsem** [comment]
i picked an input variable that will most likely not be used for anything else, we don't want accidental language changes because someone uses "locale" as a field name in a form. __elgg_lang is in line with other reserved GET variables like __elgg_ts and elgg_token.

**Comment by ewinslow** [comment]
__elgg_ts and __elgg_token should never be visible to end users.

Seems like using a nice name could be a breaking change, and I really don't want to include something with leading underscores.

I hate situations like this. :)

How about "hl"? Google uses this a lot of this purpose.

**Comment by jdalsem** [comment]
is 'hl' an abbreviation? Human Language or something? 'hl' sounds good to me. Does not seem to be very likely to be used as a parameter. Anyone else with a better option?

**Comment by mrclay** [comment]
hl is fine IMO

**Comment by jdalsem** [comment]
found out hl = host language (at Google)

**Comment by jdalsem** [comment]
Renamed the attribute. Merge when you have the time :)

**Comment by ewinslow** [comment]
Can we use a service instead of the global function? And add a test stub?

Excited to get this in but want to keep our standards high...

**Comment by jdalsem** [comment]
Added a working test and using a service now. Ready to be merged. Scrutinizer is only complaining about not receiving the code coverage data in time, no other issues.

**Comment by juho-jaakkola** [comment]
LGTM

**Comment by ewinslow** [comment]
Fantastic work, thanks so much for this!

**Pull request merged**